### PR TITLE
Tdl 20774 adding integration tests

### PR DESCRIPTION
**Pull request by rdeshmukh15**
# Description of change

- Added all the tap-tester scenarios for tap-zuora

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/test_zuora_all_fields.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+     
+    def name(self):
+        return "tap_tester_zuora_all_fields_test"
+
+    def test_run(self):
+        """ Executing tap-tester scenarios for both types of zuora APIs AQUA and REST"""
+         # Testing for only AQuA mode to reduce the execution time
+        self.run_test("AQUA")
+
+    def run_test(self, api_type):
```
Comment (RushiT0122) [inline tests/test_zuora_all_fields.py:22]: We can simply define `api_type` inside and keep only `test_run()` method unless I am missing something.

File: tests/test_zuora_all_fields.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            with self.subTest(stream=stream):
+
+                # Expected values
+                expected_all_keys = stream_to_all_catalog_fields[stream]
+                expected_automatic_keys = expected_automatic_fields.get(stream, set())
+
+                # Verify that more than just the automatic fields are replicated for each stream.
+                self.assertTrue(expected_automatic_keys.issubset(
+                    expected_all_keys), msg='{} is not in "expected_all_keys"'.format(expected_automatic_keys-expected_all_keys))
+
```
Comment (RushiT0122) [inline tests/test_zuora_all_fields.py:69]: Suggestion:
`self.assertTrue(expected_automatic_keys.issubset(expected_all_keys),
                 msg='{} is not in "expected_all_keys"'.format(expected_automatic_keys-expected_all_keys))`

File: tests/test_zuora_automatic_fields.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                record_messages_keys = [set(row.get('data').keys()) for row in data.get('messages', {})]
+                primary_keys_list = [tuple(message.get('data', {}).get(expected_pk) for expected_pk in expected_primary_keys)
+                                    for message in data.get('messages', [])
+                                    if message.get('action') == 'upsert']
+                unique_primary_keys_list = set(primary_keys_list)
+
+                # Verify that you get some records for each stream
+                self.assertGreater(
+                    record_count_by_stream.get(stream, -1), 0,
+                    msg="The number of records is not over the stream max limit for the {} stream".format(stream))
```
Comment (RushiT0122) [inline tests/test_zuora_automatic_fields.py:64]: We don't need this assert explicitly since it is taken care by `record_count_by_stream = self.run_and_verify_sync(conn_id)` here https://github.com/singer-io/tap-zuora/blob/TDL-20774-adding-integration-tests/tests/base.py#L350

File: tests/test_zuora_bookmarking.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    @staticmethod
+    def convert_state_to_utc(date_str):
+        """
+        Convert a saved bookmark value of the form '2020-08-25T13:17:36-07:00' to
+        a string formatted utc datetime,
+        in order to compare aginast json formatted datetime values
+        """
+        date_object = dateutil.parser.parse(date_str)
+        date_object_utc = date_object.astimezone(tz=pytz.UTC)
+        return datetime.datetime.strftime(date_object_utc, "%Y-%m-%dT%H:%M:%SZ")
```
Comment (RushiT0122) [inline tests/test_zuora_bookmarking.py:25]: This already defined in the `base.py`

File: tests/test_zuora_bookmarking.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            state_as_datetime = dateutil.parser.parse(state)
+            calculated_state_as_datetime = state_as_datetime - datetime.timedelta(days=days, hours=hours, minutes=minutes)
+            # convert back to string and format
+            calculated_state = datetime.datetime.strftime(calculated_state_as_datetime, "%Y-%m-%dT%H:%M:%S.000000Z")
+            stream_to_calculated_state[stream] = calculated_state
+            bookmark[repl_key[0]] = ""
+            bookmark[repl_key[0]] = calculated_state
+      
+        return stream_to_calculated_state["bookmarks"]
+
```
Comment (RushiT0122) [inline tests/test_zuora_bookmarking.py:53]: This already defined in the base.py

File: tests/test_zuora_discovery.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                                msg="There is NOT only one top level breadcrumb for {}".format(stream) + \
+                                "\nstream_properties | {}".format(stream_properties))
+
+                actual_fields = []
+                for md_entry in metadata:
+                    if md_entry['breadcrumb'] != []:
+                        actual_fields.append(md_entry['breadcrumb'][1])
+
+                # Verify there are no duplicate/conflicting metadata entries.
+                self.assertEqual(len(actual_fields), len(set(actual_fields)), msg = "duplicates in the metadata entries retrieved")
```
Comment (RushiT0122) [inline tests/test_zuora_discovery.py:88]: Break this long line.

Comment (RushiT0122) [reply tests/test_zuora_discovery.py:88]: Also mention duplicate field names in the exception (if any).

File: tests/test_zuora_discovery.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+                # verify replication key(s)
+                self.assertSetEqual(expected_replication_keys, actual_replication_keys)
+
+                # verify primary key(s)
+                self.assertSetEqual(expected_primary_keys, actual_primary_keys)
+
+                # verify the actual replication matches our expected replication method
+                self.assertEqual(expected_replication_method, actual_replication_method) # BUG_TDL-9711
```
Comment (RushiT0122) [inline tests/test_zuora_discovery.py:97]: `# BUG_TDL-9711` looks unrelated, can be removed.

File: tests/test_zuora_discovery.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        # verify the tap only discovers the expected streams
+        found_catalog_names = {catalog['tap_stream_id'] for catalog in found_catalogs}
+        self.assertSetEqual(streams_to_test, found_catalog_names)
+        LOGGER.info("discovered schemas are OK")
+
+        for stream in streams_to_test:
+            with self.subTest(stream=stream):
+                catalog = next(iter([catalog for catalog in found_catalogs
+                                     if catalog["stream_name"] == stream]))
+                assert catalog  # based on previous tests this should always be found
```
Comment (RushiT0122) [inline tests/test_zuora_discovery.py:52]: Use unittest assertion `self.assertIsNotNone()`.

File: tests/test_zuora_interrupted_sync.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        full_sync_state = menagerie.get_state(conn_id)
+
+        # State to run 2nd sync
+        #   orders: currently syncing
+        #   charges: synced records successfully
+        #   subscriptions: remaining to sync
+        state = {
+            "current_stream": "OrderAction",
+            "bookmarks":{'RevenueEventItem': {'version': 1665422431, 'UpdatedDate': dt.strftime(utils.now() - timedelta(days=5), "%Y-%m-%dT00:00:00Z")}}
+        }
```
Comment (RushiT0122) [inline tests/test_zuora_interrupted_sync.py:71]: Format dict to reduce the line length.

Comment (RushiT0122) [reply tests/test_zuora_interrupted_sync.py:71]: On the contrary, IMO, we can simply use `calculated_states_by_stream()` method in `base.py` to calculate the interrupted sync.

File: tests/test_zuora_interrupted_sync.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+                        for record in interrupted_records:
+                            rec_time = dt.strptime(record.get(expected_replication_key), "%Y-%m-%dT%H:%M:%S.%fZ")
+                            self.assertGreaterEqual(rec_time, interrupted_bookmark_datetime)
+
+                            self.assertIn(record, full_records, msg='Incremental table record in interrupted sync not found in full sync')
+
+                        # Record count for all streams of interrupted sync match expectations
+                        full_records_after_interrupted_bookmark = 0
+                        for record in full_records:
+                            rec_time = dt.strptime(record.get(expected_replication_key), "%Y-%m-%dT%H:%M:%S.%fZ")
```
Comment (RushiT0122) [inline tests/test_zuora_interrupted_sync.py:135]: Change variable name

File: tests/test_zuora_all_fields.py (hunk 1 above)
Comment (rdeshmukh15) [reply tests/test_zuora_all_fields.py:22]: Initially, executing the run_test() for both the types of APIs `REST` and `AQUA` but removed it for `REST` afterwords to reduce the execution time. Wil keep it as it is to maintain the consistency throughout the integration test cases.

File: tests/test_zuora_automatic_fields.py (hunk 3 above)
Comment (rdeshmukh15) [reply tests/test_zuora_automatic_fields.py:64]: Made the changes

File: tests/test_zuora_bookmarking.py [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+                    self.assertIsNotNone(first_bookmark_key_value)
+                    self.assertIsNotNone(first_bookmark_value)
+
+                    # Verify the second sync sets a bookmark of the expected form
+                    self.assertIsNotNone(second_bookmark_key_value)
+                    self.assertIsNotNone(second_bookmark_value)
+
+                    # Verify the second sync bookmark is Greater than or Equal to the first sync bookmark
+                    # as data changes during test
+                    self.assertGreaterEqual(second_bookmark_value, first_bookmark_value)
```
[the remark this answers is not in the document]
Comment (rdeshmukh15) [reply tests/test_zuora_bookmarking.py:160]: Tried with `assertGreater(second_bookmark_value, first_bookmark_value)` and observed that data changes during the test execution and hence, modified it to `assertGreaterEqual`

File: tests/test_zuora_discovery.py (hunk 8 above)
Comment (rdeshmukh15) [reply tests/test_zuora_discovery.py:52]: made the changes

File: tests/test_zuora_discovery.py (hunk 7 above)
Comment (rdeshmukh15) [reply tests/test_zuora_discovery.py:97]: removed

File: tests/test_zuora_discovery.py (hunk 6 above)
Comment (rdeshmukh15) [reply tests/test_zuora_discovery.py:88]: done

File: tests/test_zuora_interrupted_sync.py (hunk 10 above)
Comment (rdeshmukh15) [reply tests/test_zuora_interrupted_sync.py:135]: done

File: tests/test_zuora_interrupted_sync.py (hunk 9 above)
Comment (rdeshmukh15) [reply tests/test_zuora_interrupted_sync.py]: made the changes

File: tests/test_zuora_interrupted_sync.py [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+                    interrupted_bookmark_value = interrupted_sync_states['bookmarks'][stream]
+                    if stream in completed_streams:
+                        # Verify at least 1 record was replicated in the second sync
+                        self.assertGreaterEqual(second_sync_count,
+                                            1, 
+                                            msg="Incorrect bookmarking for {0}, at least one or more record should be replicated".format(stream))
+
+                    elif stream == interrupted_sync_states.get('current_stream', None):
+                        # For interrupted stream records sync count should be less equals
+                        self.assertLessEqual(second_sync_count,
```
Comment (RushiT0122) [inline tests/test_zuora_interrupted_sync.py:128]: Just a suggestion, here we can count the number of remaining records and have assertion like `self.assertGreaterEqual(second_sync_count, remaining_records)`.

File: tests/test_zuora_discovery.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+                                "\nstream_properties | {}".format(stream_properties))
+
+                actual_fields = []
+                for md_entry in metadata:
+                    if md_entry['breadcrumb'] != []:
+                        actual_fields.append(md_entry['breadcrumb'][1])
+
+                # Verify there are no duplicate/conflicting metadata entries.
+                self.assertEqual(len(actual_fields), len(set(actual_fields)), \
+                    msg = "duplicates in the metadata entries retrieved")
```
Comment (RushiT0122) [inline tests/test_zuora_discovery.py:91]:
```suggestion
                self.assertEqual(len(actual_fields), len(set(actual_fields)), \
                    msg = "Found following duplicate keys in the metadata entries retrieved: {set([x for x in actual_fields if actual_fields.count(x) > 1])}")
```

File: tests/test_zuora_discovery.py (hunk 7 above)
Comment (RushiT0122) [reply tests/test_zuora_discovery.py]: It is still there on L#104 and L#106

**Review by RushiT0122 — APPROVED** [review]

**Review by luandy64 — APPROVED** [review]
